src/azure-cli/azure/cli/command_modules/aro/tests/latest/custom_preparers.py

```diff
@@ -0,0 +1,127 @@
+        super(AROClusterServicePrincipalPreparer, self).__init__(
+            name_prefix, 
+            skip_assignment,
+            parameter_name,
+            parameter_password,
+            dev_setting_sp_name,
+            dev_setting_sp_password,
+            key,
+        )
+        self.client_id_to_replace = None
+        self.client_secret_to_replace = None
+        self.delete_app_on_completion = False
+
+    def create_resource(self, name, **kwargs):
+        client_id, client_secret = self._get_csp_credentials(name)
+
+        self.test_class_instance.kwargs[self.key] = client_id
+        self.test_class_instance.kwargs["{}_pass".format(self.key)] = client_secret
+
+        return {
+            self.parameter_name: client_id,
+            self.parameter_password: client_secret,
+        }
+    
+    # Overriden because RoleBasedServicePrincipal.remove_resource does not delete 
+    # the underlying AAD application generated when creating the service principal
+    def remove_resource(self, name, **kwargs):
+        super().remove_resource(name, **kwargs)
+
+        if not self.dev_setting_sp_name:
+            self.live_only_execute(self.cli_ctx, 'az ad app delete --id {}'.format(self.result.get('appId')))
+
+    def process_request(self, request):
+        if self.client_id_to_replace in request.uri:
+            request.uri = request.uri.replace(self.client_id_to_replace, MOCK_GUID)
+
+        if is_text_payload(request) and isinstance(request.body, bytes):
+            request.body = self._replace_byte_keys(request.body)
+        elif is_text_payload(request) and isinstance(request.body, str):
+            request.body = self._replace_string_keys(request.body)
+
+        return request
+
+    def process_response(self, response):
+        if is_text_payload(response) and response['body']['string']:
+            response['body']['string'] = self._replace_string_keys(response['body']['string'])
+
+        return response
+    
+    
+    def _get_csp_credentials(self, name):
+        if self.live_test or self.test_class_instance.in_recording:
+            client_id, client_secret = self._generate_csp(name)
+
+            # call AbstractPreparer.moniker to make resource counts and self.resource_moniker consistent between live
+            # and play-back. see SingleValueReplacer.process_request, AbstractPreparer.__call__._preparer_wrapper
+            # and ScenarioTest.create_random_name. This is so that when self.create_random_name is called for the
+            # first time during live or playback, it would have the same value.
+            # In short, the default sp preparer in live mode does not call moniker, which leads to inconsistent counts.
+            _ = self.moniker
+
+            self.client_id_to_replace = client_id
+            self.client_secret_to_replace = client_secret
+
+        else:
+            client_id = MOCK_GUID
+            client_secret = MOCK_SECRET
+
+        return client_id, client_secret
```

Comment:
Not a big problem but we can even remove the else now (we can replace highlighted text with)
```python
            return client_id, client_secret # this inside the if
       return MOCK_GUID, MOCK_SECRET # this outside the if

        
```

Reply:
Will do. I'll be inverting the if statement at the top of the function as well to "un-nest" the majority of the function body.